driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,247 @@
+                        "Skipping batch {} because it has already been handled.",
+                        batch_id.0
+                    );
+                    self.wait_for_next_batch_id(batch_id);
+                }
+                Ok((batch_id, Action::NotEnoughTimeLeft)) => {
+                    info!(
+                        "Skipping batch {} because there is not enough time left.",
+                        batch_id.0
+                    );
+                    self.metrics.auction_skipped(batch_id.0.into());
+                    self.wait_for_next_batch_id(batch_id);
+                }
+                Ok((batch_id, Action::Solve)) => {
+                    let deadline = batch_id.start_time() + self.batch_wait_time;
+                    if let Ok(duration) = deadline.duration_since(now) {
+                        info!(
+                            "Waiting {}s to handle batch {}.",
+                            duration.as_secs(),
+                            batch_id.0
+                        );
+                        self.wait_for_batch_id(batch_id, Instant::now() + duration);
+                    }
+                    info!("Running solver for batch {}.", batch_id.0);
+                    if let Err(err) = self.driver.run(batch_id.0.into()) {
+                        error!("StableXDriver error: {}", err);
+                    }
+                    self.past_auctions.insert(batch_id);
+                    self.wait_for_next_batch_id(batch_id);
+                }
+                Err(err) => {
+                    error!("Scheduler error: {}", err);
+                    thread::sleep(Duration::from_secs(10));
+                }
+            }
+        }
+    }
+
+    /// Return current batch id and what to do with it.
+    fn determine_action(&self, now: SystemTime) -> Result<(BatchId, Action)> {
+        let solving_batch = BatchId::currently_being_solved(now)
+            .with_context(|| anyhow!("failed to get batch id currently being solved"))?;
+        // unwrap here because this cannot fail because the `solving_batch`'s
+        // start time is always before `now`.
+        let elapsed_time = now.duration_since(solving_batch.start_time()).unwrap() - BATCH_DURATION;
+
+        Ok((
+            solving_batch,
+            if self.past_auctions.contains(&solving_batch) {
+                Action::AlreadyHandled
+            } else if elapsed_time > self.max_batch_elapsed_time {
+                Action::NotEnoughTimeLeft
+            } else {
+                Action::Solve
+            },
+        ))
+    }
+
+    /// Wait until the batch id matches the batch id according to the order book
+    /// but at most until deadline.
+    fn wait_for_batch_id(&self, batch_id: BatchId, deadline: Instant) {
+        info!("Waiting for the next batch ({}) to begin.", batch_id.0);
+        let sleep_interval = Duration::from_secs(5);
+        loop {
+            if let Ok(index) = self.orderbook_reader.get_auction_index() {
```

Comment:
What is your though on having the dual logic "either new batch mined or deadline expired"? It adds a bunch of complexity and one could argue that only having the deadline makes the code more predictable and less prone to 1 block reorgs that could happen.

Reply:
I agree.